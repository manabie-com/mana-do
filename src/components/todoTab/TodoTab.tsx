import { Button } from "components/button";
import { DATA_BUTTONS } from "components/todoTab/data";
import { EnhanceTodoStatus } from "containers/ToDoPage";
import React, { Dispatch, FC, memo } from "react";
import "./styles.scss";

export interface ITodoTabProps {
  selectedStatus: EnhanceTodoStatus;
  onClickSelectStatus: Dispatch<React.SetStateAction<EnhanceTodoStatus>>;
  onDeleteAllTodo: () => void;
}

const TodoTab: FC<ITodoTabProps> = ({
  selectedStatus,
  onClickSelectStatus,
  onDeleteAllTodo,
  ...props
}) => {
  const handleClickSelectStatus = (status: EnhanceTodoStatus): void => {
    onClickSelectStatus(status);
  };

  const handleDeleteAlldoto = (): void => {
    if (onDeleteAllTodo) onDeleteAllTodo();
  };

  return (
    <>
      <div className="todo-tab" {...props} data-testid="todo-tab">
        {DATA_BUTTONS.map(({ id, title, status }) => (
          <Button
            key={id}
            aria-label={title}
            onClick={() => handleClickSelectStatus(status)}
            className={`btn-todo--main ${
              selectedStatus === status ? "btn-todo--selected" : ""
            }`}
          >
            {title}
          </Button>
        ))}
      </div>
      <Button
        aria-label="Clear all"
        onClick={handleDeleteAlldoto}
        className="btn-todo--danger"
      >
        Clear all todos
      </Button>
    </>
  );
};

export default memo(TodoTab);
