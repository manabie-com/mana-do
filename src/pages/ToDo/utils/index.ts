import { Todo, TodoStatus } from '../../../models/todo';
const created_date = new Date().toISOString();
export const defaultToDos = {
  content: 'Content',
  created_date,
  status: TodoStatus.ACTIVE,
  id: 'testId',
  user_id: 'firstUser',
} as Todo;

export function spawnToDos(count: number): Todo[] {
  return new Array(count).fill(defaultToDos).map(todo => {
    todo.id = `testId-${Math.random()}`;
    return todo;
  });
}
