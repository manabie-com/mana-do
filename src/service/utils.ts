import {Todo, TodoStatus} from "../models/todo";

export interface TResponse {
    status: Number,
    code: string,
    data?: any
}

export const updateTodoService = async (todoId:string, content: string): Promise<TResponse> => {

    const dataListTodo:any = await localStorage.getItem('todolist') || '[]';

    const dataToAPI = await JSON.parse(dataListTodo) as Todo[]

    const index = dataToAPI.findIndex(todo => todo.id === todoId)

    dataToAPI[index].content = content

    await localStorage.setItem('todolist', JSON.stringify(dataToAPI))

    return {status: 200, code: "success"}

}

export const removeTodoService = async (todoId: string): Promise<TResponse> => {
    const dataListTodo:any = await localStorage.getItem('todolist') || '[]';

    const dataToAPI = await JSON.parse(dataListTodo) as Todo[]

    const index = dataToAPI.findIndex(todo => todo.id === todoId)

    dataToAPI.splice(index, 1)

    await localStorage.setItem('todolist', JSON.stringify(dataToAPI))

    return {status: 200, code: "success"}

}

export const updateStatusTodoService = async (todoId:string, status: TodoStatus): Promise<TResponse> => {

    const dataListTodo:any = await localStorage.getItem('todolist') || '[]';

    const dataToAPI = await JSON.parse(dataListTodo) as Todo[]

    const index = dataToAPI.findIndex(todo => todo.id === todoId)

    dataToAPI[index].status = status

    await localStorage.setItem('todolist', JSON.stringify(dataToAPI))

    return {status: 200, code: "success"}

}