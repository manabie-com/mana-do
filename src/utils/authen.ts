import {getTokenStorage} from './storeageUtils';

export const isAuthenticated = () => Boolean(getTokenStorage());
