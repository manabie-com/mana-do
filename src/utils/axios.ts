import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
const axiosClient = axios.create({
  baseURL: 'https://61615cd137492500176312cc.mockapi.io/',
  headers: {
    'Content-Type': 'application/json',
  },
});
// Add a request interceptor
axiosClient.interceptors.request.use(
  function (config: AxiosRequestConfig) {
    // Do something before request is sent
    return config;
  },
  function (error) {
    // Do something with request error
    return Promise.reject(error);
  }
);

// Add a response interceptor
axiosClient.interceptors.response.use(
  function (response: AxiosResponse) {
    // Any status code that lie within the range of 2xx cause this function to trigger
    // Do something with response data
    return response;
  },
  function (error) {
    // Any status codes that falls outside the range of 2xx cause this function to trigger
    // Do something with response error
    return Promise.reject(error);
  }
);
export default axiosClient;