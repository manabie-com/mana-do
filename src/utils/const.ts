export const TO_DO_LIST = "TO_DO_LIST";