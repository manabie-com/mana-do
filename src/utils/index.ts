import { TodoStatus } from "models";

export function isTodoCompleted(status: TodoStatus | undefined): boolean {
  return status === TodoStatus.COMPLETED;
}

export function isTodoActive(status: TodoStatus | undefined): boolean {
  return status === TodoStatus.ACTIVE;
}

export function changeTodoStatus(checked: boolean): TodoStatus {
  return checked ? TodoStatus.COMPLETED : TodoStatus.ACTIVE;
}

export function getTodoStatus(checked: boolean): TodoStatus {
  return checked ? TodoStatus.COMPLETED : TodoStatus.ACTIVE;
}
