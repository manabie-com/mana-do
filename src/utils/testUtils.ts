import { screen } from "@testing-library/react";

export function getElementByTestId(name: string): HTMLElement {
  return screen.getByTestId(name);
}

export function getAllElementByTestId(name: string): HTMLElement[] {
  return screen.getAllByTestId(name);
}

export function getElementByRole(role: string): HTMLElement {
  return screen.getByRole(role);
}
