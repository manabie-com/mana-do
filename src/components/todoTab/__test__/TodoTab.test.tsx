import { render, RenderResult, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { getElementByTestId } from "utils/testUtils";
import TodoTab, { ITodoTabProps } from "../TodoTab";

const onClickSelectStatusAction = jest.fn();
const onDeleteAllTodoAction = jest.fn();

const renderComponent = ({
  selectedStatus,
  onClickSelectStatus,
  onDeleteAllTodo,
  ...props
}: ITodoTabProps): RenderResult => {
  return render(
    <TodoTab
      selectedStatus={selectedStatus}
      onClickSelectStatus={onClickSelectStatus}
      onDeleteAllTodo={onDeleteAllTodo}
      {...props}
    />
  );
};

describe("<TodoItem /> rendering", () => {
  it("should match snapshot", () => {
    const wrapper = renderComponent({
      selectedStatus: "ALL",
      onClickSelectStatus: onClickSelectStatusAction,
      onDeleteAllTodo: onDeleteAllTodoAction,
    });
    expect(wrapper.container).toMatchSnapshot();
  });

  it("should render correcly", () => {
    renderComponent({
      selectedStatus: "ALL",
      onClickSelectStatus: onClickSelectStatusAction,
      onDeleteAllTodo: onDeleteAllTodoAction,
    });
    const element = getElementByTestId("todo-tab");
    const buttonElements = element.querySelectorAll("button.btn-todo");

    const clearAllButtonElement = screen.getByRole("button", {
      name: /Clear all/i,
    });

    expect(clearAllButtonElement).toBeInTheDocument();
    expect(clearAllButtonElement).toHaveClass("btn-todo--danger");

    expect(element).toBeInTheDocument();
    expect(buttonElements).toHaveLength(3);
  });

  it("should render button with selected class", () => {
    renderComponent({
      selectedStatus: "ALL",
      onClickSelectStatus: onClickSelectStatusAction,
      onDeleteAllTodo: onDeleteAllTodoAction,
    });
    const buttonElement = screen.getByRole("button", { name: "ALL" });
    expect(buttonElement).toHaveClass("btn-todo--selected");
  });
});

describe("<TodoItem /> interacting", () => {
  it("should handle onClick when clicked", () => {
    renderComponent({
      selectedStatus: "ALL",
      onClickSelectStatus: onClickSelectStatusAction,
      onDeleteAllTodo: onDeleteAllTodoAction,
    });
    const buttonElement = screen.getByRole("button", { name: /ACTIVE/i });
    userEvent.click(buttonElement);
    expect(onClickSelectStatusAction).toHaveBeenCalled();
  });

  it("should handle onClick when clear all button was clicked", () => {
    renderComponent({
      selectedStatus: "ALL",
      onClickSelectStatus: onClickSelectStatusAction,
      onDeleteAllTodo: onDeleteAllTodoAction,
    });
    const clearAllButtonElement = screen.getByRole("button", {
      name: /Clear all/i,
    });
    userEvent.click(clearAllButtonElement);
    expect(onDeleteAllTodoAction).toHaveBeenCalled();
  });
});
