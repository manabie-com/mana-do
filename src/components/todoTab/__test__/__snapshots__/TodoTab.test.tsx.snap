// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoItem /> rendering should match snapshot 1`] = `
<div>
  <div
    class="todo-tab"
    data-testid="todo-tab"
  >
    <button
      aria-label="ALL"
      class="btn-todo btn-todo--main btn-todo--selected"
      data-testid="button"
    >
      ALL
    </button>
    <button
      aria-label="ACTIVE"
      class="btn-todo btn-todo--main "
      data-testid="button"
    >
      ACTIVE
    </button>
    <button
      aria-label="COMPLETED"
      class="btn-todo btn-todo--main "
      data-testid="button"
    >
      COMPLETED
    </button>
  </div>
  <button
    aria-label="Clear all"
    class="btn-todo btn-todo--danger"
    data-testid="button"
  >
    Clear all todos
  </button>
</div>
`;
